### PR TITLE
Implement ERC6909 for balance accounting

### PR DESCRIPTION
**Pull request by zhongeric**
## Related Issue
We've decided to merge in ERC6909 support for now to unblock integrators and further development on core. However, we have certain optimizations that we want to make to the underlying spec, and are working closely with @jtriley-eth and others. 

We hope to move towards a token standard in the future for these types of user balances so this PR and any changes made in it should not be considered final and will very likely be changed in the future.

## Description of changes
- Uses #358 , removes claims logic
- Uses current implementation of ERC6909 in solmate with totalSupply removed
- Adds some gas tests

### CODE REVIEW

File: src/PoolManager.sol
Comment (hensha256) [inline src/PoolManager.sol:20]: I'd rather we copy and paste this contract in, and rewrite the `_mint` and `_burn` functions to not use `totalSupply`, end remove the `totalSupply` mapping entirely. It feels wrong that we still have that mapping but its always 0. And I think it'll be neater if the PoolManager can just call `_mint` and _`burn` instead of having the logic locally

File: src/PoolManager.sol
Comment (hensha256) [inline src/PoolManager.sol:348]: can you put `settle` back up before `mint` and `burn` please? Unless theres a reason to change the order? The diff is harder to review

File: test/PoolManager.t.sol
Comment (hensha256) [inline test/PoolManager.t.sol:443]: mmm maybe we should ask the EIP to change this to Mint and Burn. Its such a waste of gas of burn to always emit address(0), and same with burn

Comment (hensha256) [reply test/PoolManager.t.sol:443]: thoughts?

**Review by hensha256 — CHANGES_REQUESTED** [review]

File: src/PoolManager.sol
Comment (hensha256) [reply src/PoolManager.sol:20]: oh i see @jtriley-eth is doing a PR to do this within solmate itself... i guess we will hold this PR til thats done

Comment (snreynolds) [reply src/PoolManager.sol:20]: Kind of agree. So either we hold this PR until its completed in solmate or copy/paste/credit but modify totalSupply and then later use the solmate package when its ready. Kind of in favor of the latter so we can unblock other PRs for core.

File: test/PoolManager.t.sol
Comment (snreynolds) [reply test/PoolManager.t.sol:443]: I agree this is wasteful :/

File: test/Claims.t.sol
Comment (snreynolds) [inline test/Claims.t.sol:11]: Can we keep some of these same tests/modify them for 6909?

**Review by snreynolds — CHANGES_REQUESTED** [review]

File: test/Claims.t.sol
Comment (zhongeric) [reply test/Claims.t.sol:11]: We could if we're rolling our own implementation but I don't really think there's a need to have these tests if they're already in the 6909 repo?

File: src/PoolManager.sol
Comment (zhongeric) [reply src/PoolManager.sol:20]: Sure I can copy and paste the solmate impl into here

File: src/PoolManager.sol
Comment (zhongeric) [reply src/PoolManager.sol:348]: Sure!

File: test/PoolManager.t.sol
Comment (zhongeric) [reply test/PoolManager.t.sol:443]: Resolved offline that we're ok with Transfer

File: test/Claims.t.sol
Comment (zhongeric) [reply test/Claims.t.sol:11]: Using solmate's impl

File: src/PoolManager.sol
Comment (zhongeric) [reply src/PoolManager.sol]: Using solmate impl

File: src/PoolManager.sol
Comment (marktoda) [inline src/PoolManager.sol:345]: now thaat we are 6909, wonder if we should go back to `uint256 tokenId` as the first param here? Bit weird to be able to do `mint(currency, to, amount)` but not be able to do `transfer(currency, to, amount)`

**Review by marktoda — COMMENTED** [review]
Nice, EZ and looks great! 

Seems we probably need some tests for standard 6909 functionality as well though, to ensure no interactions and functions as expected. Also good to have full core functionality tested locally to ensure i.e. upstream changes to solmate don't affect our understandings and invariants in core. i.e. claim transfers, operators, approvals etc

File: src/PoolManager.sol
Comment (zhongeric) [reply src/PoolManager.sol:345]: Yea I think it is a nice benefit that all funcs on core use Currency

Comment (zhongeric) [reply src/PoolManager.sol:345]: Oh hmm, dealing with merge conflicts and it definitely is annoying that `balanceOf` requires uint while mint/burn use Currency.

File: src/PoolManager.sol
Comment (hensha256) [inline src/PoolManager.sol:390]: i think id rather this be in a helper contract - feels weird that PM is having to deal with allowances and operators when thats a token-specific thing
maybe we could have an intermediate contract like `V4Core6909` or something that is `6909` but adds `burnFrom` logic?

File: src/PoolManager.sol
Comment (hensha256) [reply src/PoolManager.sol:386]: i kind of love a bit of overloading lol, but i know we're keeping v4 core minimal

Comment (snreynolds) [reply src/PoolManager.sol:386]: I do think an ordering change would be nice though. So (receiver, id/currency, amount) to mirror 6909?

**Review by snreynolds — COMMENTED** [review]
Looks great. I think also just updating CurrencyLib so that we have balanceOf and transfer with native Currency type for underlying erc20 and 6909 makes sense. But lmk thoughts.

File: src/PoolManager.sol
Comment (zhongeric) [reply src/PoolManager.sol:390]: Yea agree, was kinda thinking that too. Will implement

File: src/V46909.sol
Comment (zhongeric) [inline src/V46909.sol:11]: internal

File: src/PoolManager.sol
Comment (zhongeric) [reply src/PoolManager.sol]: discussed offline: reordered parameters and using Currency for mint/burn + uint256 for other 6909 specific functions

File: test/PoolManager.t.sol
Comment (snreynolds) [inline test/PoolManager.t.sol:528]: this should be vm.expectEmit() which does true for everything

File: test/PoolManager.t.sol
Comment (snreynolds) [inline test/PoolManager.t.sol:560]: same here I think we want vm.expectEmit()

File: test/PoolManager.t.sol
Comment (snreynolds) [inline test/PoolManager.t.sol:545]: same

File: test/PoolManager.t.sol
Comment (snreynolds) [inline test/PoolManager.t.sol:577]: same

File: test/PoolManager.t.sol
Comment (snreynolds) [inline test/PoolManager.t.sol:592]: and again :)

**Review by snreynolds — COMMENTED** [review]
One small change.. I may have missed calling out some of the vm.expectEmit() s but any event check before the transfer event should check all the topics and the data field so I think you can do vm.expectEmit()

**Review by snreynolds — DISMISSED** [review]

File: src/V46909.sol
Comment (hensha256) [inline src/V46909.sol:16]: wondering if it would be nicer to have a custom error here?
like `if (amount > senderAllowance) revert X()` ?

i guess solmate doesnt have one on transferFrom though

Comment (snreynolds) [reply src/V46909.sol:16]: non standard for transferFrom flows i think ..

File: test/V46909.t.sol
Comment (hensha256) [inline test/V46909.t.sol:19]:
```suggestion
    function test_burnFrom_withApproval(address sender, Currency currency, uint256 mintAmount, uint256 transferAmount)
```

File: test/V46909.t.sol
Comment (hensha256) [inline test/V46909.t.sol:40]: please can you add a test that checks that it reverts when it doesnt have approval?

File: src/V46909.sol
Comment (marktoda) [reply src/V46909.sol:6]: also nit but I think `abstract contract` is more appropriate here as it's not intended to be deployed standalone

**Review by hensha256 — COMMENTED** [review]
just a last couple of things

File: src/PoolManager.sol
Comment (marktoda) [inline src/PoolManager.sol:380]: why switch the ordering here?

**Review by marktoda — COMMENTED** [review]
looks great, just some minor suggestions

File: src/PoolManager.sol
Comment (snreynolds) [reply src/PoolManager.sol:380]: this matches the underlying 6909 mint/burn functions

File: src/V46909.sol
Comment (zhongeric) [reply src/V46909.sol:16]: yeah not sure about this one, seems like solmate / ERC6909 generally relies on underflow / overflow checks instead of explicitly throwing

File: test/V46909.t.sol
Comment (zhongeric) [reply test/V46909.t.sol:40]: this is likely caught by the fuzz test right? i.e. when `mintAmount` is 0 the approval won't go through and thus all calls would revert?

File: src/V46909.sol
Comment (zhongeric) [reply src/V46909.sol:6]: yeah not opposed, I do prefer `ERC6909Claims` over V46909 lol

Comment (zhongeric) [reply src/V46909.sol]: Changed to ERC6909Claims

File: src/interfaces/IPoolManager.sol
Comment (zhongeric) [reply src/interfaces/IPoolManager.sol:13]: Hmm yeah so it's a bit weird because the solmate implementation doesn't use the interface either, so getting a ton of compile errors due to conflicting implementations

Comment (zhongeric) [reply src/interfaces/IPoolManager.sol]: Resolved offline - going to copy in the solmate implementation locally and inherit the interface

File: test/V46909.t.sol
Comment (zhongeric) [reply test/V46909.t.sol]: Added solmate tests which catch this

File: src/interfaces/IPoolManager.sol
Comment (snreynolds) [inline src/interfaces/IPoolManager.sol:184]: Hmm what about placing these in the IERC6909Claims interface? Just thinking now that especially we've changed the interface to match the underlying 6909 one maybe it should? 

Could go either way tho..

File: test/utils/Deployers.sol
Comment (snreynolds) [inline test/utils/Deployers.sol:10]: unnecessary import?

File: test/ERC6909Claims.t.sol
Comment (snreynolds) [inline test/ERC6909Claims.t.sol:17]: maybe just add one with no approval on burn from? make sure it reverts

File: src/interfaces/IPoolManager.sol
Comment (zhongeric) [reply src/interfaces/IPoolManager.sol:184]: Eh I think they should be in PoolManager since they are wrappers around the internal _mint and _burn and since they have the delta accounting logic in them

File: test/PoolManager.t.sol
Comment (hensha256) [inline test/PoolManager.t.sol:517]: `test_swap_mint6909IfNativeOutputNotTaken_gas`

**Review by hensha256 — DISMISSED** [review]
lgtm!

**Review by snreynolds — APPROVED** [review]
looks good!

**Comment by zhongeric** [comment]
Closing in favor of #447

**Pull request closed without merging**
